### PR TITLE
Starcoder2 model - bis

### PR DESCRIPTION
**Pull request by RaymondLi0**
The Starcoder2 model, adapted from Mistral. ~~All changes are done through options, so Mistral itself is still supported.~~ 
Main changes:

- Use layer norm
- Use standard MLP
- Add biases (optional)
- Change (default?) tokenizer class
*Embedding and residual dropout

It does not support absolute embeddings, so can't support Santacoder or Starcoder

Starcoder2-3B model: https://huggingface.co/bigcode/starcoder2-3b

Todo:

- Forward changes from https://github.com/huggingface/transformers/pull/27931, https://github.com/huggingface/transformers/pull/29027 (and future changes from Feb. 19) (in a future PR?)
- Documentation
- Copyright
- Point to starcoder2 checkpoint
- Other minor things (see todos)

@younesbelkada @ArthurZucker @jlamypoirier

### CODE REVIEW

**Review by ArthurZucker — COMMENTED** [review]
In #29228 I mention that static cache is not a blocker for the PR 😉

**Review by younesbelkada — APPROVED** [review]
Thanks ! LGTM once we add the docs !

File: docs/source/en/model_doc/starcoder2.md
Comment (ArthurZucker) [inline docs/source/en/model_doc/starcoder2.md:23]: this is fairly short. Let's add that the main difference with mistral is dropout, as the authors would be nice to explain how much this influenced training for example.

**Review by ArthurZucker — APPROVED** [review]
Let's go!

**Comment by ArthurZucker** [comment]
Thank you for the PR @RaymondLi0. cc @loubnabnl and @lvwerra, next time let's make sure doc and paper links are fully completed before we merge, we require this for every model / organisation regardless of the release date 😉 let's make this an exception

**Pull request merged**
